packages/vscode-extension/src/utilities/extensionContext.ts

```diff
@@ -15,15 +19,187 @@ export const extensionContext = new Proxy<ExtensionContext>({} as ExtensionConte
+        if (index > -1) {
+          this.onChangeAppRootListeners.splice(index, 1);
+        }
+      },
+    };
+  }
+
+  getAppRoot(): string {
+    if (!this.appRoot) {
+      throw new Error("App root not set.");
+    }
+    return this.appRoot;
+  }
+
+  setAppRoot(newAppRoot: string): void {
+    this.appRoot = newAppRoot;
+    this.onChangeAppRootListeners.forEach((listener) => {
+      listener(newAppRoot);
+    });
+    Logger.debug(`App root was set to: ${this.appRoot}.`);
+  }
 }
 
-export function getAppRootFolder() {
-  if (!_appRootFolder) {
-    throw new Error("App root folder has not been set");
+export const getCurrentLaunchConfig = (): LaunchConfigurationOptions => {
+  const launchConfiguration = workspace.getConfiguration(
+    "launch",
+    workspace.workspaceFolders![0].uri
+  );
+
+  const configurations = launchConfiguration.get<Array<Record<string, any>>>("configurations")!;
+
+  const RNIDEConfiguration = configurations.find(
+    ({ type }) => type === "react-native-ide" || type === "radon-ide" // for compatibility we want to support old configuration type name
+  );
+
+  if (!RNIDEConfiguration) {
+    return {};
   }
-  return _appRootFolder;
+
+  const { android, appRoot, ios, isExpo, metroConfigPath, env } = RNIDEConfiguration;
+
+  return { android, appRoot, ios, isExpo, metroConfigPath, env };
+};
+
+export function findAppRootCandidates(maxSearchDepth: number = 3): string[] {
+  const candidates: string[] = [];
+  const searchedFileNames = [
+    "metro.config.js",
+    "metro.config.ts",
+    "app.json",
+    "app.config.js",
+    "app.config.ts",
+  ];
+
+  // In order to optimize the search time we exclude directories,
+  // that shouldn't contain applications.
+  const excludedDirectoryPatterns: RegExp[] = [/^node_modules$/, /^ios$/, /^android$/, /^\..+/];
+
+  const searchQueue: [string, number][] | undefined = workspace.workspaceFolders?.map(
+    (workspaceFolder) => {
+      return [workspaceFolder.uri.path, 0];
+    }
+  );
+
+  if (searchQueue === undefined) {
+    Logger.warn("[FindFiles] Could not determine active workspace");
+    return [];
+  }
+
+  let currentDir: [string, number] | undefined = searchQueue.shift();
+  while (currentDir !== undefined) {
+    if (currentDir[1] > maxSearchDepth) {
+      break;
+    }
+
+    const filesAndDirs = fs.readdirSync(currentDir[0].toString(), { withFileTypes: true });
+
+    let matched = false;
+
+    filesAndDirs.forEach((dirEntry) => {
+      if (dirEntry.isFile()) {
+        if (!matched && searchedFileNames.includes(dirEntry.name)) {
+          candidates.push(currentDir![0]);
+          matched = true;
+        }
+        return;
+      }
+
+      if (excludedDirectoryPatterns.some((pattern) => pattern.test(dirEntry.name))) {
+        return;
+      }
+
+      searchQueue.push([currentDir![0] + "/" + dirEntry.name, currentDir![1] + 1]);
+    });
+    currentDir = searchQueue.shift();
+  }
```

Comment:
I think the body of this `forEach` callback is pretty complex.
I'd rewrite it into two array operations instead, one for checking if you're the candidate and another for selecting children to search. Something along the
```suggestion
    const isCandidate = filesAndDirs.some(dirEntry => {
      return dirEntry.isFile() && searchedFileNames.includes(dirEntry.name);
    });
    
    if (isCandidate) { candidates.push(currentDir[0]) }
    
    const childrenToSearch = filesAndDirs.filter(dirEntry => {
      return !excludedDirectoryPatterns.some((pattern) => pattern.test(dirEntry.name))
    }).map(dirEntry => [currentDir![0] + "/" + dirEntry.name, currentDir![1] + 1]);
    searchQueue.push(...childrenToSearch);

    currentDir = searchQueue.shift();
  }
```

Reply:
The point of creating this function was to optimize and iterators in js are not lazy, so this is not a zero-cost abstraction, I agree this would be more readable, but It is a worthy trade of in my opinion